src/qt/walletmodel.cpp

```diff
@@ -1462,18 +1491,51 @@ int WalletModel::getDefaultConfirmTarget() const
     return nTxConfirmTarget;
 }
 
-void WalletModel::handleBip47Keys(int receiverAccountNum)
+void WalletModel::handleBip47Keys(int receiverAccountNum, void * pBlockIndex_)
 {
+    //These statics are to display only one password prompt at a time and block consequent prompts
+    static std::mutex singlePasswordPromptMutex, queueMutex;
+    static std::deque<bip47::CAccountReceiver const *> receiverAccountNumQueue;
+
+    if(!pBlockIndex_)
+        return;
+    CBlockIndex * pBlockIndex = reinterpret_cast<CBlockIndex *>(pBlockIndex_);
+
     if (wallet->GetBip47Wallet()) {
         bip47::CAccountReceiver const * acc = wallet->GetBip47Wallet()->getReceivingAccount(uint32_t(receiverAccountNum));
         if (!acc)
             return;
-        UnlockContext ctx(requestUnlock(tr("Please unlock your wallet to receive a <b>RAP/BIP47 transaction</b>.")));
-        if(!ctx.isValid()) {
-            QMessageBox::critical(0, tr("RAP tx received"),
-                tr("BIP47 protocol requires unlocking your wallet every time a RAP tx is received."));
+
+        std::unique_lock<std::mutex> _(singlePasswordPromptMutex, std::try_to_lock);
+        if (!_.owns_lock()) {
+            std::lock_guard<std::mutex> _(queueMutex);
+            receiverAccountNumQueue.push_back(acc);
             return;
         }
-        bip47::utils::AddReceiverSecretAddresses(*acc, *wallet);
+
+        static QString const unlockText = tr("You have received a payment to a RAP address, please unlock your wallet to receive.");
+        UnlockContext ctx(requestUnlock(unlockText));
+        while(!ctx.isValid()) {
+            CancelPasswordDialog msgDialog(
+                    tr("RAP address payment"),
+                    tr("RAP addresses require you to unlock your wallet every time a payment to it is received."),
+                    3
+            );
+            msgDialog.setInformativeText(tr("If you do not enter your password now, you will need to rescan your wallet to receive your FIRO.<br/><br/>Re-enter your password?"));
+            if(msgDialog.exec() == QMessageBox::Cancel)
+                return;
+            ctx = requestUnlock(unlockText);
+        }
+        ctx.delayRelock(60);
+        {
+            std::lock_guard<std::mutex> _(queueMutex);
+            for(std::deque<bip47::CAccountReceiver const *>::iterator iter = receiverAccountNumQueue.begin(); iter != receiverAccountNumQueue.end(); ++iter)
+                bip47::utils::AddReceiverSecretAddresses(**iter, *wallet);
+            receiverAccountNumQueue.clear();
+        }
+        LOCK(cs_main);
+        if (pBlockIndex != chainActive.Tip()) {
+            wallet->ScanForWalletTransactions(pBlockIndex, false, false);
```

Comment:
1. do we need to check if reorg happened and `pBlockIndex` is not on the chain anymore? (`chainActive.Contains`?)
2. why start with this [presumably already scanned] block and not the next one (`chainActive.Next(pBlockIndex)`?)

Reply:
1. This is possible theoretically, so fixed, please review
2. There can be other bip47 transactions in this block that's why we unlock wallet and rescan all the blocks from this one to the tip